### ISSUE TITLE
MIDI Output: Events for NoteOn/NoteOff sent out in wrong order

### ISSUE THREAD

**Issue by SDwarfs** [LMMS/lmms#4437]
I noticed that NoteOff/NoteOff-Events are sent in the wrong order. This is an issue if a note with the same MIDI-Note-Number is played multiple times without a gap (e.g. a D3 1/2-note is played 2 times within a bar).

In that case you'll see the following Midi Events:
TIME: 0 -- NoteOn: C4, Velocity: 64
TIME: 2 -- NoteOn: C4, Velocity: 64
TIME: 2 -- NoteOff: C4
TIME: 4 -- NoteOff: C4

This is like the keyboards key for C4 is pressed, and then pressed again (after a half bar), without being actually released and then released immediately and again released (after another half bar).
This can cause issues on some MIDI devices (1st note can be "hanging" in a bad implementation) and the second note is actually stopped immediately, when it is started, and then stopped again, when it actually should be. 

This instead it should be:
TIME: 0 -- NoteOn: C4, Velocity: 64
TIME: 2 -- NoteOff: C4
TIME: 2 -- NoteOn: C4, Velocity: 64
TIME: 4 -- NoteOff: C4

The issue is, that LMMS has NoteOn/NoteOff events for the same Note and sends them  That means: LMMS should look for NoteOff-Events first, send them... and then look for NoteOn-Events and send them afterwards. If there is any sorted event queuing mechanism implemented, either the sorting criteria should be adapted (1st prio: timestamp, 2nd prio: NoteOff -> NoteOn) -- or there should be two seperated queues for NoteOff and NoteOn events which are then processed in correct order.

This happens with LMMS 1.1.3 and LMMS 1.20-rc6 when using the tool LoopBe1 (might also be causing the issue, but I've no way to find out right now) and SuperCollider's MidiClient, which works fine with a usual MIDI-Keyboard as input, but not with LMMS sending its data via LoopBe1 to it.

**Comment by zonkmachine** [comment]
Could this be the cause of https://github.com/LMMS/lmms/issues/3297 ?

**Comment by SDwarfs** [comment]
Nope, the issue seems not related to that. The MIDI Pattern from #3297 for some reason survives the obove described test (internal midi-loop). In the pattern from #3297 there is a little gap between the notes, which is visible if you use a 400%-zoom. However, if I replace the three notes by three full length 1/32th notes, only the first note is played correctly... And using LMMS 1.2.0-rc6 this doesn't cause any issues with my SF2-Player using the Arachno SoundFont.
Anyway, the issue might have a similar cause. So this might have been fixed for the internal NoteOn/NoteOff-handling, but has not been fixed in the Midi-Output-Code... I had a quick look in the source code and the behaviour of NoteOn and NoteOff was handled at multiple locations. But, I could not find any suspicious code snippet, yet.

**Comment by PhysSong** [comment]
Note on event is sent in `NotePlayHandle` constructor, while note off event is sent in `NotePlayHandle::noteOff`. When playing a pattern, the constructor is called at the beginning of a mixer cycle. However, `noteOff` function is called while processing. It leads to sending MIDI events with wrong priorities.

**Comment by SDwarfs** [comment]
This seems to be a good solution. This should even work for adjacent notes (note1.endTime == note2.startTime), if the check is based on the current on/off-state for that MIDI-noteNumber on the same MIDI-channel.
However, make sure, the counter is reset, whenever the "stop" button is pressed or any other event that stops the music playing (where all the notes currently "on" are sent a NoteOff).. Else there would be this bug: Play until counter value is >= 1 (within overlapping range), stop... and press play again: the first note is not played (no NoteOn is sent), as the counter value is > 0. Maybe check, if this is handled this way for the internal Midi-based plugins too... would be a good place for the counter reset anyway.

Also, there should be a separate counter for both (internal/external MIDI), if it is handled at different places in the code. The cleanest solution would be to have a separate counter for each device that receives MIDI (internal or external; see section below). Especially, the counter should not be stored at a track-related object, as multiple tracks might send MIDI to the same MIDI-device.

I would place this code (counting mutiple-NoteONs + filtering NoteOff-events until counter is zero) directly inside the method which finally sends out the NoteOn/NoteOff-Events to the device/plugin. This is where the "limitation" of the MIDI-protocol is actually happening and where this should be handled.  And, I would implement a method like "AllNotesOff" (if not already existent)  to stop all currently running notes, which also resets the counters, which is then called whenever Music is "stopped completely".

Thanks for working on this, DomClark! :)

**Comment by zonkmachine** [comment]
> Should we do the same for outputting to external MIDI too? It should be an easy change.

Yes. I bumped this to 1.3.
